File: RobotLocomotion/data/graph/Graph.cpp
#include "Graph.hpp"
#include <cmath> // for abs
#include <iostream>
#include <sstream> // for std::stringstream
MapGraph::MapGraph(int length, int width) : length(length), width(width) {
    generateGraph();
}

void MapGraph::generateGraph() {
    for (int i = 0; i < length; ++i) {
        for (int j = 0; j < width; ++j) {
            std::pair<int, int> directions[] = {
                {0, -1}, {1, -1}, {1, 0}, {1, 1},
                {0, 1}, {-1, 1}, {-1, 0}, {-1, -1}
            };
            std::string directionNames[] = {
                "north", "northEast", "east", "southEast",
                "south", "southWest", "west", "northWest"
            };

            for (int k = 0; k < 8; ++k) {
                int x = i + directions[k].first;
                int y = j + directions[k].second;

                bool horizontalCondition = x >= 0 && x < length;
                bool verticalCondition = y >= 0 && y < width;

                if (horizontalCondition && verticalCondition) {
                    std::stringstream ssSrc, ssDst;
                    ssSrc << i << "-" << j;
                    ssDst << x << "-" << y;

                    std::string src = ssSrc.str();
                    std::string dst = ssDst.str();

                    int weight = 5;
                    if (directionNames[k] == "north" || directionNames[k] == "south")
                        weight = 3;
                    else if (directionNames[k] == "west" || directionNames[k] == "east")
                        weight = 4;

                    addEdge(src, dst, weight);
                }
            }
        }
    }
}

// Check if a node exists
bool MapGraph::doesNodeExist(const std::string& node) const {
    return graph.find(node) != graph.end();
}

// Add an edge between src and dst with a given weight
void MapGraph::addEdge(const std::string& src, const std::string& dst, int weight) {
    if (!doesNodeExist(src)) {
        graph[src] = {{dst, weight}};
    } else {
        graph[src][dst] = weight;
    }

    if (!doesNodeExist(dst)) {
        graph[dst] = {};
    }
}

// Get the distance between src and dst
int MapGraph::getDistance(const std::string& src, const std::string& dst) const {
    if (graph.find(src) != graph.end() && graph.at(src).find(dst) != graph.at(src).end()) {
        return graph.at(src).at(dst);
    }
    return -1; // Return -1 if no path exists
}

// Get the direction from src to dst
int MapGraph::getDirection(const std::string& src, const std::string& dst) const {
    int x0, y0, x1, y1;
    sscanf(src.c_str(), "%d-%d", &x0, &y0);
    sscanf(dst.c_str(), "%d-%d", &x1, &y1);

    int dx = x1 - x0;
    int dy = y1 - y0;

    if (dx == 0 && dy == 1) return 0;
    if (dx == 1 && dy == 1) return 45;
    if (dx == 1 && dy == 0) return 90;
    if (dx == 1 && dy == -1) return 135;
    if (dx == 0 && dy == -1) return 180;
    if (dx == -1 && dy == -1) return 225;
    if (dx == -1 && dy == 0) return 270;
    if (dx == -1 && dy == 1) return 315;

    return -1; // If not a valid direction
}

File: robot-web-application/src/graph.js
class Graph {
  constructor(length, width) {
    this.graph = {};
    this.generateGraph(length, width);
  }

  generateGraph(length, width) {
    this.length = length;
    this.width = width;

    for (let i = 0; i < this.length; i++)
      for (let j = 0; j < this.width; j++) {
        const nodes = {
          north: [i, j - 1],
          northEast: [i + 1, j - 1],
          east: [i + 1, j],
          southEast: [i + 1, j + 1],
          south: [i, j + 1],
          southWest: [i - 1, j + 1],
          west: [i - 1, j],
          northWest: [i - 1, j - 1],
        };

        for (const [node, coordinates] of Object.entries(nodes)) {
          const [x, y] = coordinates;
          var horizontalCondition = x >= 0 && x < this.length;
          var verticalCondition = y >= 0 && y < this.width;

          if (horizontalCondition && verticalCondition) {
            if (node === "north" || node === "south")
              this.addEdge(`${i}${j}`, `${x}${y}`, 10);
            else if (node === "west" || node === "east")
              this.addEdge(`${i}-${j}`, `${x}-${y}`, 10);
            else this.addEdge(`${i}-${j}`, `${x}-${y}`, 14);
          }
        }
      }
  }

  doesNodeExist(node) {
    return this.graph.hasOwnProperty(node);
  }

  addEdge(src, dst, weight) {
    if (!this.doesNodeExist(src)) {
      this.graph[src] = { [dst]: weight };
    } else {
      this.graph[src][dst] = weight;
    }

    if (!this.doesNodeExist(dst)) {
      this.graph[dst] = {};
    }
  }

  getDistance(src, dst) {
    return this.graph[src][dst];
  }

  getDirection(src, dst) {
    const [x0, y0] = src.split("-");
    const [x1, y1] = dst.split("-");

    const dx = Number(x1) - Number(x0);
    const dy = Number(y1) - Number(y0);

    let direction;
    if (dx === 0 && dy === 1) direction = 0;
    else if (dx === 1 && dy === 1) direction = 45;
    else if (dx === 1 && dy === 0) direction = 90;
    else if (dx === 1 && dy === -1) direction = 135;
    else if (dx === 0 && dy === -1) direction = 180;
    else if (dx === -1 && dy === -1) direction = 225;
    else if (dx === -1 && dy === 0) direction = 270;
    else if (dx === -1 && dy === 1) direction = 315;

    return direction;
  }

  dijkstra(start, end) {
    const findLowestCostNode = (costs, processed) => {
      let lowestCost = Infinity;
      let lowestCostNode = null;
      for (const node in costs) {
        const cost = costs[node];
        if (cost < lowestCost && !processed.includes(node)) {
          lowestCost = cost;
          lowestCostNode = node;
        }
      }
      return lowestCostNode;
    };

    const nodesExceptStart = Object.keys(this.graph).filter(
      (node) => node !== start
    );
    const costs = {};
    const parents = {};
    const processed = [];

    nodesExceptStart.forEach((node) => {
      costs[node] = Infinity;
      parents[node] = null;
    });

    const startNeighbors = Object.keys(this.graph[start]);
    startNeighbors.forEach((node) => {
      parents[node] = start;
      costs[node] = this.graph[start][node];
    });

    let currentNode = findLowestCostNode(costs, processed);
    while (currentNode !== null) {
      const cost = costs[currentNode];
      const neighbors = this.graph[currentNode];

      for (const n in neighbors) {
        const newCost = cost + neighbors[n];
        if (costs[n] > newCost) {
          costs[n] = newCost;
          parents[n] = currentNode;
        }
      }
      processed.push(currentNode);
      currentNode = findLowestCostNode(costs, processed);
    }

    let path = [end];
    let point = parents[end];
    while (point !== start) {
      path.unshift(point);
      point = parents[point];
    }
    path.unshift(start);
    return path;
  }

  dfs(start) {
    const stack = [start];
    const visited = new Set();
    visited.add(start);

    while (stack.length > 0) {
      const currentNode = stack.pop();
      const neighbors = this.graph[currentNode];
      for (const neighbor in neighbors) {
        if (!visited.has(neighbor)) {
          stack.push(neighbor);
          visited.add(neighbor);
        }
      }
    }
  }

  bfs(start) {
    const queue = [start];
    const visited = new Set();
    visited.add(start);

    while (queue.length > 0) {
      const currentNode = queue.shift();
      const neighbors = this.graph[currentNode];
      for (const neighbor in neighbors) {
        if (!visited.has(neighbor)) {
          queue.push(neighbor);
          visited.add(neighbor);
        }
      }
    }
  }
}

export default function dijkstra_computation(length, width, start, end) {
  const graph = new Graph(length, width);
  return graph.dijkstra(start, end);
}

